<a href="https://colab.research.google.com/github/abdalazizrashid/AI-Feynman/blob/master/notebook_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!git clone https://github.com/abdalazizrashid/AI-Feynman.git

Cloning into 'AI-Feynman'...
remote: Enumerating objects: 55, done.
remote: Counting objects: 100% (55/55), done.
remote: Compressing objects: 100% (38/38), done.
remote: Total 457 (delta 28), reused 35 (delta 16), pack-reused 402
Receiving objects: 100% (457/457), 31.31 MiB | 15.03 MiB/s, done.
Resolving deltas: 100% (276/276), done.


In [2]:
!pip install -r AI-Feynman/requirements.txt

In [14]:
!cd /content/AI-Feynman/Code/ && ./compile.sh && apt install -y parallel


Reading package lists... Done
Building dependency tree       
Reading state information... Done
parallel is already the newest version (20161222-1).
The following package was automatically installed and is no longer required:
  libnvidia-common-440
Use 'apt autoremove' to remove it.
0 upgraded, 0 newly installed, 0 to remove and 35 not upgraded.


In [4]:
import os
os.chdir("/content/AI-Feynman/Code/")
print(os.getcwd())

/content/AI-Feynman/Code


In [5]:
from S_run_aifeynman import run_aifeynman


## Original Code Example

In [ ]:
%%time
run_aifeynman("../example_data/","example1.txt",30,"14ops.txt", polyfit_deg=3, NN_epochs=400)

Checking for brute force + 

Checking for brute force * 

Checking polyfit 

Complexity  RMSE  Expression
[0.0, 30.834454019627426, '0']
[36.83243537730492, 29.12634010089305, cos(x0) + cos(x2 + x3) + 2.333740234375]
[55.42220420859321, 29.09077077952536, cos(x2 + x3) + 1.88584470748901]
[63.97127597744305, 28.649814958454094, '1.808310192134*(cos(((x3-x2)/pi)))**(-1)']
[73.2730466728654, 26.754024113174246, '0.2*x0**2 - 0.444444444444444*x0*x1 + 0.111111111111111*x0 + 0.2*x1**2 + 0.125*x1 + 0.2*x2**2 - 0.444444444444444*x2*x3 + 0.166666666666667*x2 + 0.2*x3**2 + 0.166666666666667*x3 + 0.2']
Checking for brute force + 

Checking for brute force * 

Checking polyfit 

Complexity  RMSE  Expression
[0.0, 30.834454019627426, '0']
[36.83243537730492, 29.12634010089305, cos(x0) + cos(x2 + x3) + 2.333740234375]
[46.5605178945111, 27.13368508267342, 1.03779721260071]
[46.56056711257602, 27.133459108345154, 1.03783261803319]
[46.56057649641146, 27.13342067603215, 1.03783936851191]
[46.560576570

array([['30.834454019627426', '4.946471395876568', '593576.5675051882',
        '0.0', '30.834454019627426', '0'],
       ['29.661717905473818', '4.890530251072885', '586863.6301287463',
        '1.0', '29.661717905473818', '1'],
       ['29.304496932666545', '4.873050165836775', '584766.019900413',
        '3.0', '29.304496932666545', '1.50000000000000'],
       ['29.66164889019033', '4.820780475275042', '578493.657033005',
        '4.0', '28.26178092441696', 'exp(-23.140666108580+exp(pi))'],
       ['0.0', '-inf', '-inf', '13.169925001442312', '0.0',
        '((x0-x1)**2 + (x2-x3)**2)**0.5']], dtype='<U32')

## Download the dataset from the main website
https://www.dropbox.com/s/7kgfr00qpokgz8w/Feynman_with_units.tar.gz?dl=0
https://www.dropbox.com/s/9i05v6yw1kbkup3/Feynman_without_units.tar.gz?dl=0

The link below is no persistant you need to go to the main dropbox download link
and inspect the page under network tab pick the `file?_download_id=xxxxxxxx` and copy it as a 
cURL command and paste down below.

In [6]:
!cd /content && curl 'https://uc4b304c377d76cb47fbccc918e2.dl.dropboxusercontent.com/cd/0/get/A860-WwfdpJE63yyVv70rwlAAmIJB8K8vnHiq-jv6a45vXzyK9a-6Yqb5zKfPHI50Znd3yw1rwwnGvoCMl2pLaOQm4N9gML1KP83d2OcgANC-g/file?_download_id=31784720549261183848105032197111167178761710119256631792641873289&_notify_domain=www.dropbox.com&dl=1' \
  -H 'authority: uc4b304c377d76cb47fbccc918e2.dl.dropboxusercontent.com' \
  -H 'upgrade-insecure-requests: 1' \
  -H 'dnt: 1' \
  -H 'user-agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.89 Safari/537.36' \
  -H 'accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9' \
  -H 'sec-fetch-site: cross-site' \
  -H 'sec-fetch-mode: navigate' \
  -H 'sec-fetch-dest: iframe' \
  -H 'referer: https://www.dropbox.com/' \
  -H 'accept-language: en-US,en;q=0.9,ar;q=0.8,ru;q=0.7' \
  --compressed > with_units.gz

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1661M  100 1661M    0     0  17.0M      0  0:01:37  0:01:37 --:--:-- 17.6M


In [7]:
!cd /content && tar -xzf with_units.gz && cd /content/AI-Feynman/Code

## Solver configurations 

In [ ]:
_CFG = {
  "dataset_path"      : "/content/Feynman_with_units",
  "operations_file"   : "./14ops.txt",
  "polynomial_degree" : 3,
  "number_of_epochs"  : 500,
  "bruteforce_time"   : 20,
  "test_percentage"   : 0,
}

In [46]:
import logging
import argparse
import pathlib
import os



from threading import active_count
from multiprocessing import Pool
from multiprocessing.pool import ThreadPool
from random import shuffle
from tabulate import tabulate
from pathlib import Path
from functools import partial


from S_run_aifeynman import run_aifeynman

_CFG = {
  "dataset_path"      : "../../Feynman_without_units/",
  "operations_file"   : "./14ops.txt",
  "polynomial_degree" : 3,
  "number_of_epochs"  : 500,
  "bruteforce_time"   : 60,
  "test_percentage"   : 0,
}

class RunAll:
    """
    Run the solver on the whole dataset
    """

    def __init__(self, *,  cfg=_CFG):
        logging.basicConfig(filename="output_no_units_parallel.log", level=logging.DEBUG)
        self.cfg = cfg
        self.results = {}

 
    def print_results(self):
        table = []
        for file, sol in self.results.items():
            table.append(sol[-1])
        print(tabulate(
                table,
                headers=[
                    "Average error",
                    "Cumulative error",
                    "Error",
                    "Symbolic expression",
                ],
            )
        )
    
    def run_solver(self, dirs=None):
        if not dirs:
            path = Path(self.cfg["dataset_path"])
            dirs = list(path.iterdir())
            shuffle(dirs) # Shuffle to sample a different file each time
            
        else:
            path=Path(self.cfg["dataset_path"])
            child = dirs
            
            

        print(f"Process PID: {os.getpid()} ---------------- Number of threads: {active_count()}" )
        self.results[str(child).split("/")[-1]] = run_aifeynman(
            pathdir=str(path.resolve()) + "/",
            filename=str(child).split("/")[-1],
            BF_try_time=int(self.cfg["bruteforce_time"]),
            BF_ops_file_type=Path(self.cfg["operations_file"]),
            polyfit_deg=int(self.cfg["polynomial_degree"]),
            NN_epochs=int(self.cfg["number_of_epochs"]),
            vars_name=[],
            test_percentage=int(self.cfg["test_percentage"]),
        )

        logging.info(self.results)
        print("@"*120)
        print("@"*120)

        self.print_results()



In [47]:
def get_files(path: str, chunks=None):
    dirs = list(Path(path).iterdir())
    dirs = [file.name for file in dirs if not (str(file).endswith("test") or str(file).endswith("train"))]
    if chunks:
      for i in range(0, len(dirs), chunks):
          yield dirs[i : i + chunks]
    yield dirs

files = list(get_files(Path("/content/Feynman_without_units")))

In [48]:
%%time

solver = RunAll().run_solver
path = Path(_CFG["dataset_path"])
#dirs = list(path.iterdir())
#chunked_dirs = list(get_files(dirs, chunks=24))
# print(chunked_dirs[0], len(chunked_dirs[0]))
#    for dd in chunked_dirs:
#       pool = Pool(len(dd))
#      print(dd, len(dd))
#    pool.map(print, dd)
#     pool.map(solver, dd)
# pool.close()

solver(files[-1][-1])


Process PID: 119 ---------------- Number of threads: 4
Checking for brute force + 

Checking for brute force * 

Checking polyfit 

Complexity  RMSE  Expression
Checking for symmetry 
 III.7.38_train


TypeError: ignored

In [16]:
!parallel

Academic tradition requires you to cite works you base your article on.
When using programs that use GNU Parallel to process data for publication
please cite:

  O. Tange (2011): GNU Parallel - The Command-Line Power Tool,
  ;login: The USENIX Magazine, February 2011:42-47.

This helps funding further development; AND IT WON'T COST YOU A CENT.
If you pay 10000 EUR you should feel free to use GNU Parallel without citing.

To silence this citation notice: run 'parallel --citation'.

parallel: Warning: Input is read from the terminal. You either know what you
parallel: Warning: are doing (in which case: YOU ARE AWESOME!) or you forgot
parallel: Warning: ::: or :::: or to pipe data into parallel. If so
parallel: Warning: consider going through the tutorial: man parallel_tutorial
parallel: Warning: Press CTRL-D to exit.


In [ ]:
%%time
RunAll(cfg=_CFG)
    

/content/Feynman_with_units/I.34.8
Checking for brute force + 

Checking for brute force * 

Checking polyfit 

Complexity  RMSE  Expression
[17.509775004326936, 0.0, '1.000000000000*(x2*(x1*(x0/x3)))']
Checking for brute force + 

Checking for brute force * 

Checking polyfit 

Complexity  RMSE  Expression
[17.509775004326936, 0.0, '1.000000000000*(x2*(x1*(x0/x3)))']
Checking for brute force + 

Checking for brute force * 

Checking polyfit 

Complexity  RMSE  Expression
[16.194602975157967, nan, 'acos(-x0 - x1 - x2 + 2)']
[17.509775004326936, 0.0, '1.000000000000*(x2*(x1*(x0/x3)))']
Checking for brute force + 

Checking for brute force * 

Checking polyfit 

Complexity  RMSE  Expression
[16.194602975157967, nan, 'acos(-x0 - x1 - x2 + 2)']
[17.509775004326936, 0.0, '1.000000000000*(x2*(x1*(x0/x3)))']
Checking for brute force + 

Checking for brute force * 

Checking polyfit 

Complexity  RMSE  Expression
[16.194602975157967, nan, 'acos(-x0 - x1 - x2 + 2)']
[17.509775004326936, 0.0, '1

## Problem: **I.34.8**
Took 4 hours and 35 minute to come up with the correct result:
$$
\omega = \dfrac{x_0 x_1 x_2}{x_3}
$$
And the correct answer is:
$$
\omega = \dfrac{qvB}{p}
$$


## Problem: **III.15.14**
Took more than 5 hours to come up with this result:
$$
m = -\tan{(\dfrac{1}{ \times x_0^{-0.93}\times x_1^{-1.03}\times x_2^{-0.469}} - 1.01)}
$$
While the actual answer is:
$$
m = \dfrac{h^2}{2Ed^2}
$$

Output sample

 Complexity  RMSE  Expression
[24.236446955124386, 32.19379398225521, 'x0*x1 + x0*x2 + x1*x2 - x1 - x2**2 + 1']
[48.33584961180911, 31.9862227670943, '0.001136817047*exp((x1+x0))']
[48.444908703757726, 31.457560514093935, '-0.957879029931+x1']
[48.6933025167606, 31.016339707140748, '0.168570230758*exp(x1)']
[51.261202734823684, 29.67151999967164, '0.999530090276*(x1*x0)']
[62.98013903114244, 29.63651158757859, '0.055232500314+((x0+log(x2))*(x1-1))']
[106.30525426555863, 29.531475422563073, x0*(x1 + log(0.408610771597111*x2)) + 0.525839149951935]
[108.02150104349855, 29.461005247973745, (x0 + log(x2))*(x1 - 0.719529628753662) - 0.0555874854326248]
[153.74798012440982, 29.440090352742136, '0.1*x0*x1*x2 + x0*x1 - 0.285714285714286*x0*x2**2 + 0.142857142857143*x0*x2*x3 + 0.833333333333333*x0*x2 - 0.428571428571429*x0*x3 - 0.4*x0 - 0.125*x1**2*x2 + 0.25*x1**2 - 0.3*x1*x2**2 + 0.25*x1*x2*x3 + x1*x2 - 0.5*x1*x3 - 0.666666666666667*x1 + 0.333333333333333*x2**3 - x2**2 - 0.111111111111111*x2*x3**2 + 0.285714285714286*x2*x3 - 0.428571428571429*x2 + 0.2*x3**2 - 0.2*x3 + 0.784697875310949']
[162.58246431046206, 29.325725950623077, x1*(x0 + 0.556240618228912)*sin(x2**0.482438892126083) - 1.94434630870819]
[200.02103605333068, 28.648861022890436, '0.1*x0*x1*x2 + 0.850311055415562*x0*x1 - 0.285714285714286*x0*x2**2 + 0.142857142857143*x0*x2*x3 + 0.833333333333333*x0*x2 - 0.428571428571429*x0*x3 - 0.4*x0 - 0.125*x1**2*x2 + 0.25*x1**2 - 0.3*x1*x2**2 + 0.25*x1*x2*x3 + x1*x2 - 0.5*x1*x3 - 0.666666666666667*x1 + 0.333333333333333*x2**3 - x2**2 - 0.111111111111111*x2*x3**2 + 0.285714285714286*x2*x3 - 0.428571428571429*x2 + 0.2*x3**2 - 0.2*x3 + 0.784697875310949']
[240.46710529464815, 28.54127970345586, '0.1*x0*x1*x2 + 0.850311055415562*x0*x1 - 0.285714285714286*x0*x2**2 + 0.142857142857143*x0*x2*x3 + 0.833333333333333*x0*x2 - 0.419409322578989*x0*x3 - 0.4*x0 - 0.125*x1**2*x2 + 0.25*x1**2 - 0.3*x1*x2**2 + 0.25*x1*x2*x3 + x1*x2 - 0.5*x1*x3 - 0.666666666666667*x1 + 0.333333333333333*x2**3 - x2**2 - 0.111111111111111*x2*x3**2 + 0.285714285714286*x2*x3 - 0.428571428571429*x2 + 0.2*x3**2 - 0.2*x3 + 0.784697875310949']
[451.0046845634026, 28.311589315571727, '0.1*x0*x1*x2 + 0.850311055415562*x0*x1 - 0.285714285714286*x0*x2**2 + 0.142857142857143*x0*x2*x3 + 0.827047248703045*x0*x2 - 0.419409322578989*x0*x3 - 0.4*x0 - 0.125*x1**2*x2 + 0.25*x1**2 - 0.3*x1*x2**2 + 0.25*x1*x2*x3 + x1*x2 - 0.453915765321264*x1*x3 - 0.666666666666667*x1 + 0.333333333333333*x2**3 - 0.896533056720137*x2**2 - 0.111111111111111*x2*x3**2 + 0.275863468660905*x2*x3 - 0.435642558638502*x2 + 0.2*x3**2 - 0.2*x3 + 0.784697875310949']
[765.2021193259455, 27.68961332841078, '0.0954709145222299*x0*x1*x2 + 0.850311055415562*x0*x1 - 0.281598845270346*x0*x2**2 + 0.142857142857143*x0*x2*x3 + 0.827047248703045*x0*x2 - 0.419409322578989*x0*x3 - 0.393857016024628*x0 - 0.125*x1**2*x2 + 0.25*x1**2 - 0.295615544580811*x1*x2**2 + 0.238779807136057*x1*x2*x3 + 1.08861622104865*x1*x2 - 0.453915765321264*x1*x3 - 0.650985290052454*x1 + 0.333333333333333*x2**3 - 0.0841340395633869*x2**2*x3 - 0.896533056720137*x2**2 - 0.111111111111111*x2*x3**2 + 0.275863468660905*x2*x3 - 0.435642558638502*x2 + 0.2*x3**2 - 0.2*x3 + 0.784697875310949']
[804.4179933090178, 27.670902454575725, '0.0954709145222299*x0*x1*x2 + 0.850311055415562*x0*x1 - 0.281598845270346*x0*x2**2 + 0.142857142857143*x0*x2*x3 + 0.827047248703045*x0*x2 - 0.419409322578989*x0*x3 - 0.393857016024628*x0 - 0.125*x1**2*x2 + 0.25*x1**2 - 0.295615544580811*x1*x2**2 + 0.238779807136057*x1*x2*x3 + 1.08861622104865*x1*x2 - 0.453915765321264*x1*x3 - 0.650985290052454*x1 + 0.333333333333333*x2**3 - 0.0841340395633869*x2**2*x3 - 0.896533056720137*x2**2 - 0.114928272861219*x2*x3**2 + 0.275863468660905*x2*x3 - 0.435642558638502*x2 + 0.2*x3**2 - 0.2*x3 + 0.784697875310949']
[1049.3949821424571, 27.65482709345228, '0.0954709145222299*x0*x1*x2 + 0.850311055415562*x0*x1 - 0.281598845270346*x0*x2**2 + 0.143093568988091*x0*x2*x3 + 0.827047248703045*x0*x2 - 0.419409322578989*x0*x3 - 0.393857016024628*x0 - 0.124525299070793*x1**2*x2 + 0.255889644781202*x1**2 - 0.295615544580811*x1*x2**2 + 0.238779807136057*x1*x2*x3 + 1.08861622104865*x1*x2 - 0.453915765321264*x1*x3 - 0.650985290052454*x1 + 0.332150095352896*x2**3 - 0.0841340395633869*x2**2*x3 - 0.896533056720137*x2**2 - 0.114928272861219*x2*x3**2 + 0.275863468660905*x2*x3 - 0.435642558638502*x2 + 0.19934446687829*x3**2 - 0.198171936948463*x3 + 0.784697875310949']
[1287.6255549331968, 27.57826998133378, '0.0324957167678879*x0*x1**2 + 0.0954709145222299*x0*x1*x2 - 0.0518367881947689*x0*x1*x3 + 0.850311055415562*x0*x1 - 0.281598845270346*x0*x2**2 + 0.142857142857143*x0*x2*x3 + 0.827047248703045*x0*x2 + 0.0195647972984031*x0*x3**2 - 0.419409322578989*x0*x3 - 0.393857016024628*x0 - 0.0325007929091572*x1**3 - 0.124525299070793*x1**2*x2 + 0.0649419539246933*x1**2*x3 + 0.255889644781202*x1**2 - 0.295615544580811*x1*x2**2 + 0.238779807136057*x1*x2*x3 + 1.08861622104865*x1*x2 - 0.0388670663126931*x1*x3**2 - 0.453915765321264*x1*x3 - 0.650985290052454*x1 + 0.332150095352896*x2**3 - 0.0841340395633869*x2**2*x3 - 0.896533056720137*x2**2 - 0.114928272861219*x2*x3**2 + 0.275863468660905*x2*x3 - 0.435642558638502*x2 + 0.00642744864450299*x3**3 + 0.19934446687829*x3**2 - 0.198171936948463*x3 + 0.784697875310949']
[1327.5202240795804, 27.57798240828162, '0.0324957167678879*x0*x1**2 + 0.0954709145222299*x0*x1*x2 - 0.0518367881947689*x0*x1*x3 + 0.850311055415562*x0*x1 - 0.281598845270346*x0*x2**2 + 0.143093568988091*x0*x2*x3 + 0.827047248703045*x0*x2 + 0.0195647972984031*x0*x3**2 - 0.419409322578989*x0*x3 - 0.393857016024628*x0 - 0.0325007929091572*x1**3 - 0.124525299070793*x1**2*x2 + 0.0649419539246933*x1**2*x3 + 0.255889644781202*x1**2 - 0.295615544580811*x1*x2**2 + 0.238779807136057*x1*x2*x3 + 1.08861622104865*x1*x2 - 0.0388670663126931*x1*x3**2 - 0.453915765321264*x1*x3 - 0.650985290052454*x1 + 0.332150095352896*x2**3 - 0.0841340395633869*x2**2*x3 - 0.896533056720137*x2**2 - 0.114928272861219*x2*x3**2 + 0.275863468660905*x2*x3 - 0.435642558638502*x2 + 0.00642744864450299*x3**3 + 0.19934446687829*x3**2 - 0.198171936948463*x3 + 0.784697875310949']
Checking for brute force + 